### PR TITLE
Bug fix: deadlocked worker pool threads during a controlplane connection failure.

### PR DESCRIPTION
**Pull request by VirajSalaka**
### Purpose
Bug Fix: When the retry limit reaches its maximum limit, the worker threads which calls the control plane gets blocked. 

### Issues

Fixes #

### Automation tests
 - Unit tests added: No
 - Integration tests added: No

### Tested environments

Locally tested in the following manner.
Started choreo-product-apim locally
then Started choreo-connect locally pointing to local choreo-product-apim and ASB connection string. (I used PDP mode in choreo-connect here)
Stopped the choreo-product-apim so that we could replicate a network unreachable type failure.
Sent deploy events via ASB 

Monitored the behavior with and without the fix. And with the fix it did not blocked.

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [ ] Assigned milestone to the github issue(s)

### CODE REVIEW

**Review by renuka-fernando — DISMISSED** [review]

File: adapter/internal/synchronizer/apis_fetcher.go [hunk 1]
```diff
@@ -234,8 +234,17 @@ func FetchAPIsFromControlPlane(updatedAPIID string, updatedEnvs []string, envToD
 			break
 		} else if data.ErrorCode >= 400 && data.ErrorCode < 500 {
 			logger.LoggerSync.Errorf("Error occurred when retrieving API %q from control plane: %v", updatedAPIID, data.Err)
-			health.SetControlPlaneRestAPIStatus(false)
+
+			// If API is not found (404), then there is no point in setting the control plane status as unhealthy.
+			if data.ErrorCode != 404 {
+				health.SetControlPlaneRestAPIStatus(false)
```
Comment (renuka-fernando) [inline adapter/internal/synchronizer/apis_fetcher.go:240]: Do we set this back to healthy, in somewhere, if this has happened and resolved?

Comment (VirajSalaka) [reply adapter/internal/synchronizer/apis_fetcher.go:240]: Idea here is if the status code 401, 403 , we don't retry rather we make the adapter to be killed as it is unrecoverable. That was the initial thought. But now um wondering if we do such thing in first place. May be a log alert is the correct way to move forward.

**Review by renuka-fernando — APPROVED** [review]

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : dev-deployment-v2 : 20240129.9

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : dev-deployment-v2 : 20240129.9

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : dev-deployment-v2 : 20240129.9

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : stage-deployment-v2 : 20240201.1

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : stage-deployment-v2 : 20240201.1

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : stage-deployment-v2 : 20240201.1

**Comment by choreo-cicd** [comment]
[succeeded] Controlplane cluster : prod-deployment-v2 : 20240201.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(EastUS) cluster : prod-deployment-v2 : 20240201.3

**Comment by choreo-cicd** [comment]
[succeeded] Dataplane(NorthEU) cluster : prod-deployment-v2 : 20240201.3

**Pull request merged**